File: CSharpWithLanguageExt/Account/Domain/Account.cs
using LanguageExt;

using Time = Lib.Time;
using Lib.Types;
using Bank.Transfer.Domain;

namespace Bank.Account.Domain;

using StateTransitionResult = Validation<Err, (Event Event, AccountState NewState)>;

public static class Account {
   public static string StreamName(Guid id) => $"accounts_{id}";

   public static Map<string, Type> EventTypeMapping = (
      (nameof(CreatedAccount), typeof(CreatedAccount)),
      (nameof(DebitedTransfer), typeof(DebitedTransfer)),
      (nameof(DebitedAccount), typeof(DebitedAccount)),
      (nameof(DailyDebitLimitUpdated), typeof(DailyDebitLimitUpdated)),
      (nameof(DepositedCash), typeof(DepositedCash)),
      (nameof(LockedCard), typeof(LockedCard)),
      (nameof(UnlockedCard), typeof(UnlockedCard)),
      (nameof(RegisteredInternalTransferRecipient), typeof(RegisteredInternalTransferRecipient)),
      (nameof(RegisteredDomesticTransferRecipient), typeof(RegisteredDomesticTransferRecipient)),
      (nameof(RegisteredInternationalTransferRecipient), typeof(RegisteredInternationalTransferRecipient))
   );

   public static AccountState Apply(this AccountState acc, object evt) =>
      evt switch {
         DepositedCash e
            => acc with { Balance = acc.Balance + e.DepositedAmount },

         DebitedTransfer e
            => acc with { Balance = acc.Balance - e.DebitedAmount },

         DebitedAccount e
            => acc with {
                  Balance = acc.Balance - e.DebitedAmount,
                  DailyDebitAccrued = DailyDebitAccrued(acc, e),
                  LastDebitDate = e.Timestamp
               },

         DailyDebitLimitUpdated e
            => acc with { DailyDebitLimit = e.DebitLimit },

         LockedCard _
            => acc with { Status = AccountStatus.ActiveWithLockedCard },

         UnlockedCard _
            => acc with { Status = AccountStatus.Active },

         RegisteredInternalTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               e.AccountNumber,
               e.AsTransferRecipient()
            )
         },

         RegisteredDomesticTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               $"{e.RoutingNumber}_{e.AccountNumber}",
               e.AsTransferRecipient()
            )
         },

         RegisteredInternationalTransferRecipient e => acc with {
            TransferRecipients = acc.TransferRecipients.AddOrUpdate(
               e.Identification,
               e.AsTransferRecipient()
            )
         },

         _ => acc
      };

   public static Decimal DailyDebitAccrued(AccountState state, DebitedAccount evt) {
      // When accumulating events into AccountState aggregate...
      // -> Ignore debits older than a day
      if (!Time.IsToday(evt.Timestamp)) return 0;

      if (evt.Origin == Constants.DebitOriginMaintenanceFee) return state.DailyDebitAccrued;

      // When applying a new event to the cached AccountState & the
      // last debit event did not occur today...
      // -> Ignore the cached DailyDebitAccrued
      if (!Time.IsToday(state.LastDebitDate)) return evt.DebitedAmount;

      return state.DailyDebitAccrued + evt.DebitedAmount;
   }

   public static StateTransitionResult StateTransition(
      this AccountState state,
      Command command
   ) =>
      command switch {
         TransferCmd cmd => state.Transfer(cmd),
         DebitCmd cmd => state.Debit(cmd),
         LimitDailyDebitsCmd cmd => state.LimitDailyDebits(cmd),
         DepositCashCmd cmd => state.Deposit(cmd),
         LockCardCmd cmd => state.LockCard(cmd),
         UnlockCardCmd cmd => state.UnlockCard(cmd),
         RegisterTransferRecipientCmd cmd => state.RegisterTransferRecipient(cmd),
         _ => new Err("State Transition not implemented for command.")
      };

   public static StateTransitionResult Transfer(
      this AccountState state,
      TransferCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (state.Balance - cmd.Amount < state.AllowedOverdraft)
         return Errors.InsufficientBalance;

      if (state.TransferRecipients.Find(cmd.Recipient.Identification).IsNone)
         return TransferErr.RecipientRegistrationRequired(cmd);

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult Debit(
      this AccountState state,
      DebitCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (state.Status == AccountStatus.ActiveWithLockedCard &&
          cmd.Origin != Constants.DebitOriginMaintenanceFee)
         return Errors.AccountCardLocked;

      if (state.Balance - cmd.Amount < state.AllowedOverdraft)
         return Errors.InsufficientBalance;

      if (state.DailyDebitLimit != -1 &&
         // Maintenance fee does not count toward daily debit accrual
         cmd.Origin != Constants.DebitOriginMaintenanceFee &&
         Time.IsToday(cmd.Timestamp) &&
         state.DailyDebitAccrued + cmd.Amount > state.DailyDebitLimit
      ) {
         return Errors.ExceededDailyDebitLimit(state.DailyDebitLimit);
      }

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult LimitDailyDebits(
      this AccountState state,
      LimitDailyDebitsCmd cmd
   ) {
      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult Deposit(
      this AccountState state,
      DepositCashCmd cmd
   ) {
      if (state.Status == AccountStatus.Closed)
         return Errors.AccountNotActive;

      if (cmd.Amount <= 0)
         return Errors.InvalidDepositAmount;

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult LockCard(
      this AccountState state,
      LockCardCmd cmd
   ) {
      if (state.Status != AccountStatus.Active)
         return Errors.AccountNotActive;

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult UnlockCard(
      this AccountState state,
      UnlockCardCmd cmd
   ) {
      if (state.Status != AccountStatus.ActiveWithLockedCard)
         return new Err($"Account card already unlocked: {state.Status}");

      var evt = cmd.ToEvent();
      return (evt, state.Apply(evt));
   }

   public static StateTransitionResult RegisterTransferRecipient(
      this AccountState state,
      RegisterTransferRecipientCmd cmd
   ) {
      if (state.TransferRecipients.Find(cmd.Recipient.Identification).IsSome)
         return TransferErr.RecipientAlreadyRegistered(cmd);

      var evt = cmd.ToEvent().Unwrap();
      return (evt, state.Apply(evt));
   }

   public static AccountState Create(CreatedAccount evt) =>
      new AccountState(
         FirstName: evt.FirstName,
         LastName: evt.LastName,
         EntityId: evt.EntityId,
         Currency: evt.Currency,
         Balance: evt.Balance
      );

   public static class Constants {
      public const string DebitOriginMaintenanceFee = "actor:maintenance_fee";
   }
}

public sealed record AccountState(
   Guid EntityId,
   string FirstName,
   string LastName,
   string Currency,
   AccountStatus Status = AccountStatus.Active,
   decimal Balance = 0,
   decimal AllowedOverdraft = 0,
   decimal DailyDebitLimit = -1,
   decimal DailyDebitAccrued = 0,
   DateTime LastDebitDate = default,
   Map<string, TransferRecipient> TransferRecipients = default
);

public enum AccountStatus {
   Active,
   ActiveWithLockedCard,
   Closed
}

File: AkkaIntegrated/Account/Domain/Account.fs
[<RequireQualifiedAccess>]
module Account

open Validus

open BankTypes
open Bank.Account.Domain
open Bank.Transfer.Domain
open Lib.Types
open Lib.Time

let recipientLookupKey (recipient: TransferRecipient) =
   match recipient.AccountEnvironment with
   | RecipientAccountEnvironment.Internal -> recipient.Identification
   | RecipientAccountEnvironment.Domestic ->
      $"{recipient.RoutingNumber.Value}_{recipient.Identification}"

let DailyDebitAccrued state (evt: BankEvent<DebitedAccount>) : decimal =
   // When applying a new event to the cached AccountState & the
   // last debit event did not occur today...
   // -> Ignore the cached DailyDebitAccrued
   let dailyDebitAccrued =
      if state.LastDebitDate.IsSome && IsToday state.LastDebitDate.Value then
         state.DailyDebitAccrued
      else
         0m

   // When accumulating events into AccountState aggregate...
   // -> Ignore debits older than a day
   if IsToday evt.Data.Date then
      dailyDebitAccrued + evt.Data.DebitedAmount
   else
      dailyDebitAccrued

let applyEvent (state: AccountState) (evt: AccountEvent) =
   match evt with
   | CreatedAccount e ->
      MaintenanceFee.reset {
         AccountState.empty with
            EntityId = e.EntityId
            Email = e.Data.Email
            FirstName = e.Data.FirstName
            LastName = e.Data.LastName
            Currency = e.Data.Currency
            Balance = e.Data.Balance
            Status = AccountStatus.Active
      }
   | AccountClosed _ -> {
      state with
         Status = AccountStatus.Closed
     }
   | DepositedCash e ->
      MaintenanceFee.fromDeposit
         {
            state with
               Balance = state.Balance + e.Data.DepositedAmount
         }
         e.Data.DepositedAmount
   | DebitedAccount e ->
      MaintenanceFee.fromDebit {
         state with
            Balance = state.Balance - e.Data.DebitedAmount
            DailyDebitAccrued = DailyDebitAccrued state e
            LastDebitDate = Some e.Data.Date
      }
   | MaintenanceFeeDebited e ->
      MaintenanceFee.reset {
         state with
            Balance = state.Balance - e.Data.DebitedAmount
      }
   | MaintenanceFeeSkipped _ -> MaintenanceFee.reset state
   | DailyDebitLimitUpdated e -> {
      state with
         DailyDebitLimit = e.Data.DebitLimit
     }
   | LockedCard _ -> {
      state with
         Status = AccountStatus.CardLocked
     }
   | UnlockedCard _ -> {
      state with
         Status = AccountStatus.Active
     }
   | TransferPending e ->
      MaintenanceFee.fromDebit {
         state with
            Balance = state.Balance - e.Data.DebitedAmount
      }
   | TransferApproved _ -> state
   | TransferRejected e ->
      MaintenanceFee.fromDebitReversal {
         state with
            Balance = state.Balance + e.Data.DebitedAmount
      }
   | TransferDeposited e ->
      MaintenanceFee.fromDeposit
         {
            state with
               Balance = state.Balance + e.Data.DepositedAmount
         }
         e.Data.DepositedAmount
   | InternalTransferRecipient e ->
      let recipient = e.Data.toRecipient ()

      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }
   | DomesticTransferRecipient e ->
      let recipient = e.Data.toRecipient ()

      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }
   | InternationalTransferRecipient e ->
      let recipient = e.Data.toRecipient ()
      let key = recipientLookupKey recipient

      {
         state with
            TransferRecipients = state.TransferRecipients.Add(key, recipient)
      }

module private StateTransition =
   let transitionErr (err: StateTransitionError) =
      Error <| StateTransitionError err

   let map
      (eventTransform: BankEvent<'t> -> AccountEvent)
      (state: AccountState)
      (eventValidation: ValidationResult<BankEvent<'t>>)
      =
      eventValidation
      |> Result.mapError ValidationError
      |> Result.map (fun evt ->
         let evt = eventTransform evt
         (evt, applyEvent state evt))

   let create (state: AccountState) (cmd: CreateAccountCommand) =
      if state.Status <> AccountStatus.ReadyToOpen then
         transitionErr AccountNotReadyToActivate
      else
         map CreatedAccount state <| cmd.toEvent ()

   let deposit (state: AccountState) (cmd: DepositCashCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map DepositedCash state <| cmd.toEvent ()

   let limitDailyDebits (state: AccountState) (cmd: LimitDailyDebitsCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map DailyDebitLimitUpdated state <| cmd.toEvent ()

   let lockCard (state: AccountState) (cmd: LockCardCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map LockedCard state <| cmd.toEvent ()

   let unlockCard (state: AccountState) (cmd: UnlockCardCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map UnlockedCard state <| cmd.toEvent ()

   let debit (state: AccountState) (cmd: DebitCommand) =
      if state.Status = AccountStatus.CardLocked then
         transitionErr AccountCardLocked
      elif state.Status <> AccountStatus.Active then
         transitionErr AccountNotActive
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         transitionErr <| InsufficientBalance state.Balance
      elif
         state.DailyDebitLimit <> -1m
         && IsToday cmd.Date
         && state.DailyDebitAccrued + cmd.Amount > state.DailyDebitLimit
      then
         transitionErr <| ExceededDailyDebit state.DailyDebitLimit
      else
         map DebitedAccount state <| cmd.toEvent ()

   let maintenanceFee (state: AccountState) (cmd: MaintenanceFeeCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         transitionErr <| InsufficientBalance state.Balance
      else
         map MaintenanceFeeDebited state <| cmd.toEvent ()

   let skipMaintenanceFee
      (state: AccountState)
      (cmd: SkipMaintenanceFeeCommand)
      =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map MaintenanceFeeSkipped state <| cmd.toEvent ()

   let transfer (state: AccountState) (cmd: TransferCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      elif state.Balance - cmd.Amount < state.AllowedOverdraft then
         transitionErr <| InsufficientBalance state.Balance
      elif
         not
         <| state.TransferRecipients.ContainsKey(
            recipientLookupKey cmd.Recipient
         )
      then
         transitionErr RecipientRegistrationRequired
      else
         map TransferPending state <| cmd.toEvent ()

   let approveTransfer (state: AccountState) (cmd: ApproveTransferCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map TransferApproved state <| cmd.toEvent ()

   let rejectTransfer (state: AccountState) (cmd: RejectTransferCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map TransferRejected state <| cmd.toEvent ()

   let registerTransferRecipient
      (state: AccountState)
      (cmd: RegisterTransferRecipientCommand)
      =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      elif
         state.TransferRecipients.ContainsKey cmd.Recipient.Identification
      then
         transitionErr RecipientAlreadyRegistered
      else
         match cmd.Recipient.AccountEnvironment with
         | RecipientAccountEnvironment.Internal ->
            map InternalTransferRecipient state
            <| TransferRecipientEvent.local cmd
         | RecipientAccountEnvironment.Domestic ->
            map DomesticTransferRecipient state
            <| TransferRecipientEvent.domestic cmd
         | RecipientAccountEnvironment.International ->
            map InternationalTransferRecipient state
            <| TransferRecipientEvent.international cmd

   let depositTransfer (state: AccountState) (cmd: DepositTransferCommand) =
      if not state.CanProcessTransactions then
         transitionErr AccountTransactionProcessingDisabled
      else
         map TransferDeposited state <| cmd.toEvent ()

   let closeAccount (state: AccountState) (cmd: CloseAccountCommand) =
      map AccountClosed state <| cmd.toEvent ()

let stateTransition (state: AccountState) (command: Command) =
   match box command with
   | :? CreateAccountCommand as cmd -> StateTransition.create state cmd
   | :? DepositCashCommand as cmd -> StateTransition.deposit state cmd
   | :? DebitCommand as cmd -> StateTransition.debit state cmd
   | :? MaintenanceFeeCommand as cmd -> StateTransition.maintenanceFee state cmd
   | :? SkipMaintenanceFeeCommand as cmd ->
      StateTransition.skipMaintenanceFee state cmd
   | :? LimitDailyDebitsCommand as cmd ->
      StateTransition.limitDailyDebits state cmd
   | :? LockCardCommand as cmd -> StateTransition.lockCard state cmd
   | :? UnlockCardCommand as cmd -> StateTransition.unlockCard state cmd
   | :? TransferCommand as cmd -> StateTransition.transfer state cmd
   | :? ApproveTransferCommand as cmd ->
      StateTransition.approveTransfer state cmd
   | :? RejectTransferCommand as cmd -> StateTransition.rejectTransfer state cmd
   | :? DepositTransferCommand as cmd ->
      StateTransition.depositTransfer state cmd
   | :? RegisterTransferRecipientCommand as cmd ->
      StateTransition.registerTransferRecipient state cmd
   | :? CloseAccountCommand as cmd -> StateTransition.closeAccount state cmd
